### PR TITLE
Implement Machine fix for host-only adapter

### PR DESCRIPTION
**Pull request by thieman**
@paetling This brings https://github.com/docker/machine/pull/3112 into Dusty. We ported the Machine code that gets the VM IP into Dusty to get around Machine's slow startup times. Unfortunately, that means we need to keep the underlying code in sync when big bug fixes go out.

As part of this, attempted to memoize the `VBoxManage showvminfo` call. Busting cache whenever we take the machine up/down to make it pick up VM config changes we sometimes make, e.g. changing the NIC. Will want to run this through its paces in RC, the memoize part might have broken stuff.

### CODE REVIEW

File: dusty/systems/virtualbox/__init__.py [hunk 1]
```diff
@@ -115,6 +115,7 @@ def _stop_docker_vm():
     """Stop the Dusty VM if it is not already stopped."""
     check_call_demoted(['docker-machine', 'stop', constants.VM_MACHINE_NAME], redirect_stderr=True)
 
+@memoized
 def _get_vm_config():
     return check_output_demoted(['VBoxManage', 'showvminfo', '--machinereadable', constants.VM_MACHINE_NAME]).splitlines()
```
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: why is this worth memoizing?  Is it really slow?  seems like needless speedup with costs.

File: dusty/systems/virtualbox/__init__.py [hunk 2]
```diff
@@ -199,29 +200,55 @@ def get_docker_vm_ip():
...
-            if line.startswith('inet') and not line.startswith('inet6'):
-                ip = line.split(' ')[1].split('/')[0]
-                return ip
-
-    # If all our crazy hacks up there don't work, we can still get the IP
-    # from Machine directly. The reason we don't do this normally is
-    # because it's really slow, taking around 600ms on my fast machine.
-    logging.warning('Could not get IP the fast way, falling back to Docker Machine')
-    return check_output_demoted(['docker-machine', 'ip', constants.VM_MACHINE_NAME]).rstrip()
+                return host_port
```
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: so the first one is always the one doing the local forwarding?

File: dusty/systems/virtualbox/__init__.py [hunk 3]
```diff
@@ -199,29 +200,55 @@ def get_docker_vm_ip():
...
+    raise ValueError('Could not determine localhost port for SSH forwarding')
+
+def _get_host_only_mac_address():
+    # Get the number of the host-only adapter
+    vm_config = _get_vm_config()
+    for line in vm_config:
+        if line.startswith('hostonlyadapter'):
+            adapter_number = int(line[15:16])
```
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: is there a better way to do this, [-2:]?

File: dusty/systems/virtualbox/__init__.py [hunk 4]
```diff
@@ -199,29 +200,55 @@ def get_docker_vm_ip():
...
+
+def _get_host_only_ip():
+    """Determine the host-only IP of the Dusty VM through Virtualbox and SSH
+    directly, bypassing Docker Machine. We do this because Docker Machine is
+    much slower, taking about 600ms total. We are basically doing the same
+    flow Docker Machine does in its own code."""
+    mac = _get_host_only_mac_address()
+    ip_addr = check_output_demoted(['ssh', '-o', 'StrictHostKeyChecking=no',
+                                    '-o', 'UserKnownHostsFile=/dev/null',
+                                    '-i', _vm_key_path(), '-p', _get_localhost_ssh_port(),
```
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: so we ask docker, which is listening on localhost at a port, what its ip address is?

File: dusty/systems/virtualbox/__init__.py [hunk 5]
```diff
@@ -199,29 +200,55 @@ def get_docker_vm_ip():
...
+        raise ValueError('No host-only adapter is defined for the Dusty VM')
+
+    for line in vm_config:
+        if line.startswith('macaddress{}'.format(adapter_number)):
+            return line.split('=')[1].strip('"').lower()
+    raise ValueError('Could not find MAC address for adapter number {}'.format(adapter_number))
+
+def _ip_for_mac_from_ip_addr(ip_addr, target_mac):
+    return_next_ip = False
+    for line in ip_addr.splitlines():
```
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: why does ip_addr have multiple lines?

File: dusty/systems/virtualbox/__init__.py [hunk 6]
```diff
@@ -199,29 +200,55 @@ def get_docker_vm_ip():
...
+        line = line.strip()
+        if line.startswith('link/ether'):
+            line_mac = line.split(' ')[1].replace(':', '')
+            if line_mac == target_mac:
+                return_next_ip = True
+        elif return_next_ip and line.startswith('inet') and not line.startswith('inet6'):
+            ip = line.split(' ')[1].split('/')[0]
+            return ip
```
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: generally no idea what any of this does. Seems like magic.  I see that it is similar to the stuff above but not really sure about what it is doing.

File: dusty/systems/virtualbox/__init__.py (hunk 1 above)
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: I see, you reset it in ensure_docker_vm_is_started().  So should not change without caching being busted.  Still curious about speedup. We call it 2X less right?  Not like we are pinging this like crazy?

Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: Not hitting it super hard, but it does take about 150ms each time

File: dusty/systems/virtualbox/__init__.py (hunk 3 above)
Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: The line looks like `hostonlyadapter2="vboxnet0"`

File: dusty/systems/virtualbox/__init__.py (hunk 5 above)
Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: It's the output of `ip addr show`, run it on your VM to see what it looks like

File: dusty/systems/virtualbox/__init__.py (hunk 2 above)
Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: there should only ever be one `Forwarding` entry in the Dusty VM

File: dusty/systems/virtualbox/__init__.py (hunk 4 above)
Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: No, there's a localhost port forwarded to port 22 in the VM. We can use that to SSH to the VM before we know its actual IP address.

File: dusty/systems/virtualbox/__init__.py (hunk 6 above)
Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: there's no easy way to ask a machine to give you the IP assigned to the device with a given MAC address. this is parsing the output of `ip addr show`, which contains both of these things, to determine one from the other

File: dusty/systems/virtualbox/__init__.py (hunk 3 above)
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: worthing doing len('hostonadapater') to make it explicit?  (It might not be)

Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: I'll go liberally sprinkle comments around once I can get integration to stop flaking out on me

File: dusty/systems/virtualbox/__init__.py (hunk 4 above)
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: seems like a better variable name is needed.  network_interfaces?

File: dusty/systems/virtualbox/__init__.py (hunk 5 above)
Comment (paetling) [inline dusty/systems/virtualbox/__init__.py]: seems like a better variable name is needed. network_interfaces?

Comment (thieman) [inline dusty/systems/virtualbox/__init__.py]: gonna call it `ip_addr_show` then

**Pull request merged**